### PR TITLE
ls/linear pool supports sample output type simple case/109

### PR DESCRIPTION
**Pull request by lshandross**
This PR assumes that every unique combination of task IDs for the same model has the same number of provided component samples and target component samples. Further work will remove this assumption and allow for varying numbers of samples (as well as sampling methods beyond stratified sampling), documented in the Issue #143, but we want this functionality implemented sooner so that it can be used in other work that involves ensembles of samples.

Edit: We will stick to the simplest case documented in Issue #109 instead and proceed with further cases later

### CODE REVIEW

File: tests/testthat/test-linear_pool.R [hunk 1]
```diff
@@ -409,3 +409,114 @@ test_that("(weighted) quantiles correctly calculated - lognormal family", {
...
+  reindexed_outputs <- sample_outputs |>
+    dplyr::mutate(output_type_id = paste0(.data[["model_id"]], .data[["output_type_id"]]))
+  expected_outputs <- reindexed_outputs |>
+    dplyr::filter(output_type_id %in% c("a2", "a3", "b1", "d1", "d3"))
+  for (i in 1:2) {
+    expected_outputs <- reindexed_outputs |>
+      dplyr::filter(.data[["model_id"]] %in% letters[(i + 1):4]) |>
+      dplyr::mutate(output_type_id = paste0(.data[["output_type_id"]], i)) |>
+      dplyr::bind_rows(expected_outputs)
+  }
```
Comment (elray1) [inline tests/testthat/test-linear_pool.R:505]: Where did these expectations come from?  It's not immediately clear to me why these are "correct" results or would be expected a priori.

File: tests/testthat/test-linear_pool.R [hunk 2]
```diff
@@ -409,3 +409,114 @@ test_that("(weighted) quantiles correctly calculated - lognormal family", {
...
+test_that("ensemble of samples is correctly calculated for more complex cases", {
+  sample_outputs <- expand.grid(stringsAsFactors = FALSE,
+                                model_id = letters[1:4],
+                                location = c("222", "888"),
+                                horizon = 1, #week
+                                target = "inc death",
+                                target_date = as.Date("2021-12-25"),
+                                output_type = "sample",
+                                output_type_id = 1:3,
+                                value = NA_real_)
```
Comment (elray1) [inline tests/testthat/test-linear_pool.R:473]: Reflecting my main comment above about needing to group by the compound task id set, I think we will want a more complex example here, where:
 - there is one variable with multiple levels that predictions are "joint across" (satisfied here by location)
 - the is another variable with multiple levels that predictions are not "joint across"

That will enable us to test whether sampling is done correctly by choosing sets of samples with the same `output_type_id` within the different groups for different compound units.

**Review by elray1 — CHANGES_REQUESTED** [review]
This is a good start, but I think we will need to be more careful about things like sample trajectories that are generated separately for each location.  In that kind of situation, I think we need to randomly select a target number of trajectories from each model within each location.  See the comments related to this on line 172 of R/linear_pool_samples.R and in the test file.

There are other more minor comments throughout.

File: R/linear_pool_sample.R [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+                                      weights_col_name = "weight",
+                                      task_id_cols,
+                                      output_samples) {
+  num_models <- length(unique(model_out_tbl$model_id))
+  samples_per_model <- model_out_tbl |>
+    dplyr::group_by(dplyr::across(dplyr::all_of(c("model_id", task_id_cols)))) |>
+    dplyr::summarize(provided_n_component_samples = dplyr::n()) |>
+    dplyr::ungroup() |>
+    dplyr::select("model_id", "provided_n_component_samples") |>
+    dplyr::distinct(.keep_all = TRUE) # assumes same number per task ID combo
```
[the remark this answers is not in the document]
Comment (lshandross) [reply R/linear_pool_sample.R:123]: We could add the sample output type to be one of the forecast types that are checked in the `validate_output_type_ids()` function (and this is what I was thinking of doing) to ensure that this condition is met and generate an error if not.

**Pull request closed without merging**